### PR TITLE
Add infrastructure for testing julia's LLVM passes

### PR DESCRIPTION
**Pull request by Keno**
Infrastructure pieces split out from #21158

### CODE REVIEW

File: deps/llvm.mk [hunk 1]
```diff
@@ -124,7 +104,7 @@ ifeq ($(USE_LLVM_SHLIB),1)
 # NOTE: we could also --disable-static here (on the condition we link tools
 #       against libLLVM) but there doesn't seem to be a CMake counterpart option
 LLVM_FLAGS += --enable-shared
-LLVM_CMAKE += -DLLVM_BUILD_LLVM_DYLIB:BOOL=ON
+LLVM_CMAKE += -DLLVM_BUILD_LLVM_DYLIB:BOOL=ON -DLLVM_LINK_LLVM_DYLIB:BOOL=ON
```
Comment (tkelman) [inline deps/llvm.mk]: does this need to be conditional on llvm version?

Comment (Keno) [inline deps/llvm.mk]: Looks like 3.8 was the first version with this option. However, I don't think CMake complains too loudly if you pass it an option it doesn't recognize.

Comment (vchuravy) [inline deps/llvm.mk]: 3.8 is the first version that we actually use CMAKE on so it should not be a problem, see #19054.

File: test/llvmpasses/Makefile [hunk 2]
```diff
@@ -0,0 +1,11 @@
+SRCDIR := $(abspath $(dir $(lastword $(MAKEFILE_LIST))))
+JULIAHOME := $(abspath $(SRCDIR)/../..)
+include $(JULIAHOME)/deps/Versions.make
+include $(JULIAHOME)/Make.inc
+include $(JULIAHOME)/deps/llvm-ver.make
+BUILDDIR = $(JULIAHOME)/deps/scratch
+include $(JULIAHOME)/deps/llvm-options.mk
+LLVM_SRC_DIR:=$(JULIAHOME)/deps/srccache/llvm-$(LLVM_VER)
```
Comment (vtjnash) [inline test/llvmpasses/Makefile]: These configuration options aren't really right, since they assume a very specific in-tree source build layout. Does LLVM not provide any way to run its test framework out-of-tree?

**Review by vtjnash — APPROVED** [review]
It'll be great to have this. Thanks for splitting it out!

File: test/llvmpasses/Makefile (hunk 2 above)
Comment (Keno) [inline test/llvmpasses/Makefile]: You can build it out of tree, but you need a copy of lit which is in the LLVM source dir. You can probably make it work, but might be easier for somebody who uses out-of-tree builds to fix this up after committing this.

File: deps/llvm.mk (hunk 1 above)
Comment (tkelman) [inline deps/llvm.mk]: @Keno this flag caused https://github.com/JuliaLang/julia/issues/21621 - why was this necessary?

Comment (Keno) [inline deps/llvm.mk]: so you can load libjulia into opt and use the llvm passes.
